### PR TITLE
feat: add unit tests for notify and sync exception chaining

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
1. Refactor sync_service_impl to write unit tests around sync.
2. Write unit tests for exception chaining on notify and sync services.

NOTE: The changes in sync_manager.dart (at_client/lib/src/manager/sync_manager.dart) can be ignored. The file is currently not in use. The changes are to avoid the compile time errors.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: at_client/lib/src/service/sync_service_impl.dart [hunk 1]
```diff
@@ -365,9 +386,18 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
...
+      try {
+        syncResponseJson = JsonUtils.decodeJson(DefaultResponseParser()
+            .parse(await _remoteSecondary.executeVerb(syncBuilder))
+            .response);
+      } on AtException catch (e) {
+        e.stack(AtChainedException(Intent.syncData,
+            ExceptionScenario.remoteVerbExecutionFailed, e.message));
+        _logger.severe(
+            'Exception occurred in fetching sync response : ${e.getTraceMessage()}');
+        rethrow;
```
Comment (murali-shris) [inline at_client/lib/src/service/sync_service_impl.dart]: i think we should not rethrow here.. log and continue syncing rest of the response from server

Comment (sitaram-kalluri) [reply at_client/lib/src/service/sync_service_impl.dart]: Ok, Will remove the rethrow

**Review by gkc — CHANGES_REQUESTED** [review]
As a number of methods have changed from static methods to instance methods, please can you scan through all of the repos to see if the static methods are being used anywhere? GitHub search should allow you to search at organisation level so you can search in all repos in the atsign-foundation and atsign-company organisations respectively

**Comment by sitaram-kalluri** [comment]
> As a number of methods have changed from static methods to instance methods, please can you scan through all of the repos to see if the static methods are being used anywhere? GitHub search should allow you to search at organisation level so you can search in all repos in the atsign-foundation and atsign-company organisations respectively

Searched for the methods whose static is removed in GitHub organisation level and the methods are only used in at_client and at_functional_test. Not used outside the repo anywhere.

File: at_client/lib/src/client/remote_secondary.dart [hunk 2]
```diff
@@ -167,4 +167,17 @@ class RemoteSecondary implements Secondary {
     logger.info('auto restarting monitor');
     await monitor(command, notificationCallBack, privateKey);
   }
+
+  Intent _getIntent(VerbBuilder builder) {
```
Comment (gkc) [inline at_client/lib/src/client/remote_secondary.dart:171]: Nice!

File: at_client/lib/src/manager/sync_manager.dart [hunk 3]
```diff
@@ -48,18 +50,18 @@ class SyncManager {
     if (preference.syncStrategy == SyncStrategy.scheduled && !_isScheduled) {
       _scheduleSyncTask();
     }
+    syncUtil = SyncUtil();
```
Comment (gkc) [inline at_client/lib/src/manager/sync_manager.dart]: Why not remove this line and change the declaration line above from `SyncUtil? syncUtil;` to `SyncUtil syncUtil = SyncUtil();` ?

Comment (gkc) [reply at_client/lib/src/manager/sync_manager.dart]: Then you could change all usages of `syncUtil?.` to just `syncUtil.`

**Review by gkc — CHANGES_REQUESTED** [review]
Nearly there!

File: at_client/lib/src/manager/sync_manager.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/manager/sync_manager.dart]: This changes are implemented and pushed the changes.

**Review by gkc — APPROVED** [review]
This looks good to go. However, before final merge, please do some quick sanity checking of some apps using at_client from this branch.

FYI I made a commit which reverted your dependency override for at_commons

**Pull request merged**